_posts/2019-10-23-kafka-authentication-using-oauth-2.0.md

```diff
@@ -0,0 +1,538 @@
+The problem is that it is quite complex to setup, and people often find it hard to configure.
+
+Since Kafka version 2.0.0 there is an extensible OAuth 2.0 compatible token-based mechanism available, called SASL OAUTHBEARER.
+
+We have developed extensions that provide integration with OAuth 2.0 compliant authorization servers. 
+That means that in principle you can use any OAuth 2.0 compliant authorization server to expose centrally managed users for authentication with Kafka.
+
+
+## Why OAuth 2.0?
+
+Originally developed for scenarios where web users need to give limited access to their data inside one web site to another web site, OAuth 2.0 provides at least two main benefits.
+One is that user accounts and credentials for many services are managed centrally, organization-wide, by a dedicated identity service.
+Another important benefit is that the user only ever passes credentials like username and password to that dedicated server.
+The result of authenticating is a time-limited token which the user passes around to other services, and which serves to identify user, and to establish which permissions they have.
+ 
+
+## Hands-on
+
+Let's see how this works. As our OAuth 2.0 compliant authorization server we will use a basic setup of Keycloak with TLS support.
+We'll then configure a Kafka cluster with OAuth 2.0 support, and an example Kafka producer, and Kafka consumer client applications, that will use OAuth 2.0 to authenticate. 
+
+We'll assume that you have a Kubernetes cluster available in order to explain how this works for Strimzi, but the OAuth 2.0 support can be used with non-Strimzi Kafka installations too, by using [Strimzi Kafka OAuth libraries](https://github.com/strimzi/strimzi-kafka-oauth).
+ 
+If you are using `minikube` on your local machine, make sure to start `minikube tunnel` in another terminal in order for LoadBalancer services to get their externals IPs assigned:
+
+    sudo minikube tunnel 
+
+If you're wondering, an example of the tested `minikube` configuration is to run it as:
+
+    minikube start --vm-driver hyperkit --memory=4096 --cpus=4
+
+
+## Setting up Keycloak
+
+For any kind of real-world use, all communication with the OAuth 2.0 authorization server has to be performed over a secure connection. 
+We'll configure TLS access on Keycloak to make sure we're safe from eavesdropping or man-in-the-middle attacks.
+It is also important that all the clients and services accessing Keycloak use it through the same hostname in the url.
+That is required for token validation to work properly.
+
+Let's first create a server certificate for the Keycloak server.
+Due to cross-namespace communication, we'll have to access the server from pods as `https://keycloak.keycloak:8443`.
+We'll set the certificate's common name to `keycloak`, but add `keycloak.keycloak` as a subject alternative name for certificate hostname validation to pass.
+
+```
+# create CA key
+openssl genrsa -out ca.key 2048
```

Comment:
This makes it look much more complicated than it should. But TBH I'm not sure how to make it simpler and still do the TLs part :-(.

Reply:
Right, we want it as simple as possible. Maybe we could use some script that is packaged with strimzi/kafka docker image to generate all this for us. If you could run it with docker and mount local directory, and this way have a simple CA / server cert tooling that generates for you the keys, truststores on local disk. Maybe something to add for the next version.